spec/cov/cov_spec.rb

```diff
@@ -0,0 +1,177 @@
+    end
+
+    context "when root is in the subdirectory of the project" do
+      let(:root) { absolute_path("calculator/operations") }
+
+      it "collects code coverage including operations only" do
+        subject.start
+
+        expect(calculator.add(1, 2)).to eq(3)
+        expect(calculator.subtract(1, 2)).to eq(-1)
+
+        coverage = subject.stop
+
+        expect(coverage.size).to eq(2)
+        expect(coverage.keys).to include(
+          absolute_path("calculator/operations/add.rb"),
+          absolute_path("calculator/operations/subtract.rb")
+        )
+      end
+
+      it "clears the coverage data after stopping" do
+        subject.start
+        expect(calculator.add(1, 2)).to eq(3)
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+        expect(coverage.keys).to include(absolute_path("calculator/operations/add.rb"))
+
+        subject.start
+        expect(calculator.subtract(1, 2)).to eq(-1)
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+        expect(coverage.keys).to include(absolute_path("calculator/operations/subtract.rb"))
+      end
+
+      it "does not track coverage when stopped" do
+        subject.start
+        expect(calculator.add(1, 2)).to eq(3)
+        subject.stop
+
+        expect(calculator.subtract(1, 2)).to eq(-1)
+
+        subject.start
+        expect(calculator.multiply(1, 2)).to eq(2)
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+        expect(coverage.keys).to include(absolute_path("calculator/operations/multiply.rb"))
+      end
+
+      it "does not fail if start called several times" do
+        subject.start
+        expect(calculator.add(1, 2)).to eq(3)
+
+        subject.start
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+      end
+
+      it "does not fail if stop called several times" do
+        subject.start
+        expect(calculator.add(1, 2)).to eq(3)
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+
+        expect(subject.stop).to eq({})
+      end
+
+      it "tracks coverage in mixins" do
+        subject.start
+        expect(calculator.divide(6, 3)).to eq(2)
+        coverage = subject.stop
+        expect(coverage.size).to eq(2)
+        expect(coverage.keys).to include(absolute_path("calculator/operations/divide.rb"))
+        expect(coverage.keys).to include(absolute_path("calculator/operations/helpers/calculator_logger.rb"))
+      end
+
+      context "multi threaded execution" do
+        def thread_local_cov
+          Thread.current[:datadog_ci_cov] ||= Datadog::CI::Cov.new(root: root)
+        end
+
+        it "collects coverage for each thread separately" do
+          t1 = Thread.new do
+            cov = thread_local_cov
+            cov.start
+
+            sleep 0.1
+            expect(calculator.add(1, 2)).to eq(3)
+            sleep 0.1
+            expect(calculator.multiply(1, 2)).to eq(2)
+            sleep 0.1
```

Comment:
my intention was to force a context switch so that I am sure that they are really concurrent, that is, the first thread did not finish by the time the second is started, I will try Queue approach

Reply:
I usually try to avoid the sleeps as they can create a false sense of security -- e.g. if the threads are never concurrent, the test will still pass :sweat_smile: